tokio-trace-serde/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.0.len()))?;
+        for element in self.0 {
+            seq.serialize_element(element.name())?;
+        }
+        seq.end()
+    }
+}
+
+#[derive(Debug)]
+pub struct SerializeLevel<'a>(&'a Level);
+
+impl<'a> Serialize for SerializeLevel<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if self.0 == &Level::ERROR {
+            serializer.serialize_str("ERROR")
+        } else if self.0 == &Level::WARN {
+            serializer.serialize_str("WARN")
+        } else if self.0 == &Level::INFO {
+            serializer.serialize_str("INFO")
+        } else if self.0 == &Level::DEBUG {
+            serializer.serialize_str("DEBUG")
+        } else if self.0 == &Level::TRACE {
+            serializer.serialize_str("TRACE")
+        } else {
+            unreachable!()
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SerializeId<'a>(&'a Id);
+
+impl<'a> Serialize for SerializeId<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("Id", 1)?;
+        state.serialize_field("u64", &self.0.into_u64())?;
+        state.end()
+    }
+}
+
+#[derive(Debug)]
+pub struct SerializeMetadata<'a>(&'a Metadata<'a>);
+
+impl<'a> Serialize for SerializeMetadata<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("Metadata", 9)?;
+        state.serialize_field("name", self.0.name())?;
+        state.serialize_field("target", self.0.target())?;
+        state.serialize_field("level", &SerializeLevel(self.0.level()))?;
+        state.serialize_field("module_path", &self.0.module_path())?;
+        state.serialize_field("file", &self.0.file())?;
+        state.serialize_field("line", &self.0.line())?;
+        state.serialize_field("fields", &SerializeFieldSet(self.0.fields()))?;
+        state.serialize_field("field_names", &self.0.is_span())?;
+        state.serialize_field("callsite", &self.0.is_event())?;
+        state.end()
+    }
+}
+
+/// Implements `serde::Serialize` to write `Event` data to a serializer.
+#[derive(Debug)]
+pub struct SerializeEvent<'a>(&'a Event<'a>);
+
+impl<'a> Serialize for SerializeEvent<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let serializer = serializer.serialize_struct("Event", 2)?;
+        let mut visitor = SerdeStructVisitor {
+            serializer,
+            state: Ok(()),
+        };
+        self.0.record(&mut visitor);
+        visitor.finish()
+    }
+}
+
+/// A Serde visitor to pull `Attributes` data out of a serialized stream
```

Comment:
I think this comment should be updated to reflect that we are serializing the fields in a `Record` _to_ a serde serializer, the way you did for `SerializeEvent`.